crates/nu-protocol/src/span.rs

```diff
@@ -74,76 +73,122 @@ impl<T> IntoSpanned for T {
+    pub const fn unknown() -> Self {
+        Self { start: 0, end: 0 }
     }
 
     /// Note: Only use this for test data, *not* live data, as it will point into unknown source
     /// when used in errors.
-    pub const fn test_data() -> Span {
+    pub const fn test_data() -> Self {
         Self::unknown()
     }
 
-    pub fn offset(&self, offset: usize) -> Span {
-        Span::new(self.start - offset, self.end - offset)
+    pub fn offset(&self, offset: usize) -> Self {
+        Self::new(self.start - offset, self.end - offset)
     }
 
     pub fn contains(&self, pos: usize) -> bool {
-        pos >= self.start && pos < self.end
+        self.start <= pos && pos < self.end
     }
 
-    pub fn contains_span(&self, span: Span) -> bool {
-        span.start >= self.start && span.end <= self.end
+    pub fn contains_span(&self, span: Self) -> bool {
+        self.start <= span.start && span.end <= self.end
     }
 
-    /// Point to the space just past this span, useful for missing
-    /// values
-    pub fn past(&self) -> Span {
-        Span {
+    /// Point to the space just past this span, useful for missing values
+    pub fn past(&self) -> Self {
+        Self {
             start: self.end,
             end: self.end,
         }
     }
+
+    /// Returns the minimal [`Span`] that encompasses both of the given spans.
+    ///
+    /// The two `Spans` can overlap in the middle,
+    /// but must otherwise be in order by satisfying:
+    /// - `self.start <= after.start`
+    /// - `self.end <= after.end`
+    ///
+    /// If this is not guaranteed to be the case, use [`Span::merge`] instead.
+    pub fn append(self, after: Self) -> Self {
+        debug_assert!(
+            self.start <= after.start && self.end <= after.end,
+            "Can't merge two Spans that are not in order"
+        );
+        Self {
+            start: self.start,
+            end: after.end,
+        }
+    }
+
+    /// Returns the minimal [`Span`] that encompasses both of the given spans.
+    ///
+    /// The spans need not be in order or have any relationship.
+    ///
+    /// [`Span::append`] is slightly more efficient if the spans are known to be in order.
+    pub fn merge(self, other: Self) -> Self {
+        Self {
+            start: usize::min(self.start, other.start),
+            end: usize::max(self.end, other.end),
+        }
+    }
+
+    /// Returns the minimal [`Span`] that encompasses all of the spans in the given slice.
+    ///
+    /// The spans are assumed to be in order, that is, all consecutive spans must satisfy:
+    /// - `spans[i].start <= spans[i + 1].start`
+    /// - `spans[i].end <= spans[i + 1].end`
+    ///
+    /// (Two consecutive spans can overlap as long as the above is true.)
+    ///
+    /// Use [`Span::from_unordered`] if the spans are not known to be in order.
+    pub fn concat(spans: &[Self]) -> Self {
+        // TODO: enable assert below
+        // debug_assert!(!spans.is_empty());
```

Comment:
Disabled this assert, as this would involve more intrusive changes. A similar assert was included in the old `nu_protocol::span` and it was disabled too, so we're not necessarily losing anything here.